### PR TITLE
Introduce unicode-bom rule

### LINKED ISSUE #3792: Add unicode-bom rule
> What is the problem you're trying to solve?

The developers on my team use BOM inconsistently. This clutters the diffs.

> What solution would you like to see?

A rule that requires that files always begin with the Unicode BOM character U+FEFF or that files must never begin with U+FEFF.

Inspired by https://eslint.org/docs/rules/unicode-bom

I tried to write such a rule myself, but I could not find any traces of BOM in the results of the parsing.

### PR DESCRIPTION
**Pull request by vkrol**
> Which issue, if any, is this issue related to?

Closes #3792.

> Is there anything in the PR that needs further explanation?

No, it's self explanatory.

I think it's better to add the auto-fix functionality separately, because first we need to fix the bug in stylelint that causes BOM to be cut out in any auto-fix run (https://github.com/stylelint/stylelint/issues/4228).

### CODE REVIEW

File: lib/rules/unicode-bom/__tests__/index.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+      description: "empty with BOM",
+      message: messages.rejected
+    },
+    {
+      code: "\uFEFFa{}",
+      description: "with BOM",
+      message: messages.rejected
+    }
+  ]
+});
```
Comment (evilebottnawi) [inline lib/rules/unicode-bom/__tests__/index.js]: let's add couple test with `html` and `css-in-js`

Comment (vkrol) [reply lib/rules/unicode-bom/__tests__/index.js]: Done.

File: lib/rules/unicode-bom/index.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    const validOptions = validateOptions(result, ruleName, {
+      actual: expectation,
+      possible: ["always", "never"]
+    });
+
+    if (
+      !validOptions ||
+      root.source.inline ||
+      root.source.lang === "object-literal" ||
+      root.document !== undefined
```
Comment (vkrol) [inline lib/rules/unicode-bom/index.js]: I have some doubts about this check, but I have not found another way to verify that it is not a HTML document.

File: lib/rules/unicode-bom/__tests__/index.js (hunk 1 above)
Comment (vkrol) [reply lib/rules/unicode-bom/__tests__/index.js]: Done.

File: docs/user-guide/example-config.md [hunk 3]
```diff
@@ -170,6 +170,7 @@ You might want to learn a little about [how rules are named and how they work to
     "string-no-newline": true,
     "string-quotes": "single"|"double",
     "time-min-milliseconds": int,
+    "unicode-bom": "never",
```
Comment (hudochenkov) [inline docs/user-guide/example-config.md]:
```suggestion
    "unicode-bom": "always"|"never",
```

File: lib/rules/unicode-bom/index.js (hunk 2 above)
Comment (hudochenkov) [reply lib/rules/unicode-bom/index.js]: Could you add a comment into the code explaining what this condition checks, please.

File: lib/rules/unicode-bom/index.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    if (expectation === "always" && !hasBOM) {
+      report({
+        result,
+        ruleName,
+        message: messages.expected,
+        root,
+        line: 1
+      });
+    } else if (expectation === "never" && hasBOM) {
```
Comment (hudochenkov) [inline lib/rules/unicode-bom/index.js]:
```suggestion
    }

    if (expectation === "never" && hasBOM) {
```

It will increase readability.

**Comment by vkrol** [comment]
@hudochenkov thanks! Done.

File: lib/rules/unicode-bom/README.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+Require or disallow the Unicode Byte Order Mark.
+
+## Options
+
+`string`: `"always"|"never"`
+
+### `"always"`
+
+The following pattern is considered violation:
```
Comment (ntwb) [inline lib/rules/unicode-bom/README.md]:
```suggestion
The following pattern is considered a violation:
```

File: lib/rules/unicode-bom/README.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+```css
+a {}
+```
+
+The following pattern is *not* considered violation:
```
Comment (ntwb) [inline lib/rules/unicode-bom/README.md]:
```suggestion
The following pattern is *not* considered a violation:
```

File: lib/rules/unicode-bom/README.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+### `"never"`
+
+The following pattern is considered violation:
+
+```css
+U+FEFF
+a {}
+```
+
+The following pattern is *not* considered violation:
```
Comment (ntwb) [inline lib/rules/unicode-bom/README.md]:
```suggestion
The following pattern is *not* considered a violation:
```

File: lib/rules/unicode-bom/README.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+```css
+U+FEFF
+a {}
+```
+
+### `"never"`
+
+The following pattern is considered violation:
```
Comment (ntwb) [inline lib/rules/unicode-bom/README.md]:
```suggestion
The following pattern is considered a violation:
```

**Comment by vkrol** [comment]
@ntwb thanks, fixed.
> Ideally we update all instances of this where a single example is used

I can do this for all the rules in a separate PR.

**Comment by hudochenkov** [comment]
> -   Added: `unicode-bom` rule ([#4225](https://github.com/stylelint/stylelint/pull/4225)).

**Pull request merged**